### DISCUSSION TITLE
select2 is not working on the child component and closing the modal.`

### DISCUSSION THREAD

**Discussion by anonymouse703** [livewire/livewire#4374] [1. Help]
I have to component table(parent) and form(child) when I test select2 in the table (parent) component  it's working fine but when I test it in the form then select2 is not working.

Working only in the table component

```
document.addEventListener('DOMContentLoaded', function() {
        $('#ex').select2();//table
        $('#example').select2();//form
    });

```

I tried to put this hook but still not working

```
Livewire.hook('element.initialized', (el, component) => {
            $('#ex').select2();//table
            $('#example').select2();//form
            console.log('initialized....');
        });

```
  How to initialize it every time I open and close modal or there's a changens in the dome?

Pic for reference:

Working in Table (Parent): https://ibb.co/SNrL1fW 
Not working in form modal (Child): https://ibb.co/7JsHm4s

**Comment by denizozturk** [comment]
I can't claim to be the best method, but it is the most useful method for me. 
This my usually using method, before in master layout paste this code;

The library in the example: https://developer.snapappointments.com/bootstrap-select/

_The function that build/rebuild the select library you are using. You need to create the build/rebuild function for the Select2 library_
```
let selectpicker = (type = 'load') => {
	if(type === 'load') {
		$('.selectpicker').selectpicker();
	} else {
		$('.selectpicker').selectpicker(type);
	}
}
```

_Use build/rebuild function for load and refresh states of Livewire components_
```
document.addEventListener('livewire:load', function () {
	selectpicker()
})

document.addEventListener('livewire:refresh', () => {
	selectpicker('refresh')
})
```

_Finally, you need to listen for an event to trigger this action in the class field of the livewire component_
```
window.addEventListener('refreshSelectpicker', () => {
	selectpicker('refresh')
})
```

_In class using:_
`$this->dispatchBrowserEvent('refreshSelectpicker');`

**Comment by yrizzz** [comment]
I have used and tried many ways to solve the select2 problem on Livewire v3.
I am quite happy because it can run well
here are the steps

1. Change inject_morph_markers to false (important) at config/livewire.php 
2. Follow my code 

          $(document).ready(function(){
                  $('.select2l').select2();
                  $('.select2ltags').select2({
                      placeholder: 'Input baru ? (ketik lalu enter)',
                      tags: true,
                      allowClear: true
                  });
          
                  $('.modal').on('shown.bs.modal', function(e) {
                      $(this).find('.select2l').select2({
                          dropdownParent: $(this).closest('.modal')
                      });
                      $(this).find('.select2ltags').select2({
                          placeholder: 'Input baru ? (ketik lalu enter)',
                          tags: true,
                          allowClear: true,
                          dropdownParent: $(this).closest('.modal')
                      });
                  });
          });

if you use wire:navigate , don't forget add in livewire:navigated like : 

       document.addEventListener("livewire:navigated", function() {
            $('.select2l').select2();
            $('.select2ltags').select2({
                placeholder: 'Input baru ? (ketik lalu enter)',
                tags: true,
                allowClear: true
            });
            $('.modal').on('shown.bs.modal', function(e) {
                $(this).find('.select2l').select2({
                    dropdownParent: $(this).closest('.modal')
                });
                $(this).find('.select2ltags').select2({
                    placeholder: 'Input baru ? (ketik lalu enter)',
                    tags: true,
                    allowClear: true,
                    dropdownParent: $(this).closest('.modal')
                });
             });
         });

**Comment by waqar321** [comment]
very simple,  when to apply select2, it has its own huge js code, that interfare with livewire binding,
you only need 2 solutions, 

first probelm is to get selected data in a public variable, you can get using event emit,
second problem is, after selecting data, select2 doesn't work, you can add select2 in a div and make it wire:ignore, 

    controller:

    public $selectRoles = [];
    protected $listeners = ['selectedRoles' => 'HandleselectedRoles'];
    
    public function mount(User $user)
    {
        $this->roles  = Role::pluck('title', 'id');
    }
    public function HandleselectedRoles($Roles)
    {
        $this->selectRoles = $Roles;
   }

   view:

                        <div class="form-group" wire:ignore>
                            <label for="password">Roles</label>
                            <select class="multipleRoles" placeholder='please select roles' wire:model="selectRoles" name="roles[]" id="roles" class="form-control select2" multiple="multiple" style="width: 100%;">  
                            @foreach ($roles as $key => $roles)
                                    <option value="{{ $key }}" {{ in_array($key, old('roles', [])) || (isset($user) && $user->roles->contains($key)) ? 'selected' : '' }}>
                                        {{ $roles }}
                                    </option>
                                @endforeach
                            </select>
                        </div>
   
JS: 

 $(document).ready(function () 
        {
            $('.multipleRoles').select2();

            window.initSelectCompanyDrop=()=>{
                $('.multipleRoles').select2({
                    placeholder: 'Please Select Roles',
                    // allowClear: true
                });
            }
            initSelectCompanyDrop();

            $('.multipleRoles').on('change', function(e) 
            {
                console.log($('.multipleRoles').select2("val")); 
                Livewire.emit('selectedRoles', 
                $('.multipleRoles').select2("val"));
            });
});
